File: 08.Sort/786.第k个数.cpp
// https://www.acwing.com/problem/content/description/788/
#include <assert.h>
#include <limits.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <algorithm>
#include <iostream>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
  int kth(int *A, int b, int e, int k) {
    if (b >= e) {
      return INT_MIN;
    }

    if (b + 1 == e) {
      return A[b];
    }

    const int x = A[b + ((e - b) >> 1)];
    int l = b;
    int i = b;
    int r = e - 1;

    while (i <= r) {
      if (A[i] < x)
        swap(A[l++], A[i++]);
      else if (A[i] == x)
        i++;
      else
        swap(A[r--], A[i]);
    }

    const int lcnt = l - b;
    const int mcnt = i - l;

    // check left side
    if (k < lcnt)
      return kth(A, b, l, k);
    // check right side
    if (k >= (lcnt + mcnt))
      return kth(A, i, e, k - lcnt - mcnt);
    return x;
  }

public:
  int KthNumber(int *A, int N, int k) { return kth(A, 0, N, k - 1); }
};

int N;
int K;
int A[100001];

int main(void) {
  Solution s;
  while (scanf("%d%d", &N, &K) != EOF) {
    for (int i = 0; i < N; i++) {
      scanf("%d", A + i);
    }
    printf("%d\n", s.KthNumber(A, N, K));
  }
  return 0;
}

File: 08.Sort/786.第k个数.py
# https://www.acwing.com/problem/content/description/788/
def swap(A, i, j):
  t = A[i]
  A[i] = A[j]
  A[j] = t

def kth(A, b, e, k):
  if b >= e:
    return 0

  if b + 1 >= e:
    return A[b]

  x = A[b + ((e - b) >> 1)]
  i = b
  l = b
  r = e - 1

  while i <= r:
    if A[i] < x:
      swap(A, l, i)
      i += 1
      l += 1
    elif (A[i] == x):
      i += 1
    else:
      swap(A, r, i)
      r -= 1

  lcnt = l - b
  mcnt = i - l

  if k < lcnt:
    return kth(A, b, l, k)
  if k >= (lcnt + mcnt):
    return kth(A, i, e, k - lcnt - mcnt)
  return x

def kthNumber(A, n, k):
  return kth(A, 0, n, k - 1)

s = raw_input().split()
n = int(s[0])
k = int(s[1])
s = raw_input().split()
A = [int(x) for x in s]
print kthNumber(A, n, k)


